### ISSUE TITLE
Focusing multiple windows of a process freezes the process.

### ISSUE THREAD

**Issue by crocket** [swaywm/sway#5698]
- **Sway Version:** 1.5
wlroots version: 0.11.0

- **Debug Log:**

```
00:03:21.256 [sway/commands.c:255] Handling command 'focus'
00:03:21.256 [sway/ipc-server.c:307] Sending workspace::focus event
00:03:21.256 [sway/ipc-server.c:960] Added IPC reply of type 0x80000000 to client 24 queue: { "change": "focus", "old": { "id": 6, "name": "3: <span font-family='Font Awesome 5 Free'><\/span>", "rect": { "x": 15, "y": 47, "width": 1890, "height": 986 }, "focused": false, "focus": [ 5 ], "border": "none", "current_border_width": 0, "layout": "tabbed", "orientation": "none", "percent": null, "window_rect": { "x": 0, "y": 0, "width": 0, "height": 0 }, "deco_rect": { "x": 0, "y": 0, "width": 0, "height": 0 }, "geometry": { "x": 0, "y": 0, "width": 0, "height": 0 }, "window": null, "urgent": false, "marks": [ ], "fullscreen_mode": 1, "nodes": [ { "id": 5, "name": "Alacritty", "rect": { "x": 15, "y": 78, "width": 1890, "height": 955 }, "focused": false, "focus": [ ], "border": "pixel", "current_border_width": 4, "layout": "none", "orientation": "none", "percent": 1.0, "window_rect": { "x": 4, "y": 4, "width": 1882, "height": 951 }, "deco_rect": { "x": 0, "y": 0, "width": 1890, "height": 31 }, "geometry": { "x": 0, "y": 0, "width": 800, "height": 600 }, "window": 6291460, "urgent": false, "marks": [ ], "fullscreen_mode": 0, "nodes": [ ], "floating_nodes": [ ], "sticky": false, "type": "con", "pid": 6571, "app_id": null, "visible": false, "max_render_time": 0, "shell": "xwayland", "inhibit_idle": false, "idle_inhibitors": { "user": "none", "application": "none" }, "window_properties": { "class": "Alacritty", "instance": "tmux-main", "title": "Alacritty", "transient_for": null, "window_type": "normal" } } ], "floating_nodes": [ ], "sticky": false, "num": 3, "output": "HDMI-A-1", "type": "workspace", "representation": "T[Alacritty]" }, "current": { "id": 14, "name": "2: <span font-family='Font Awesome 5 Free'><\/span>", "rect": { "x": 15, "y": 47, "width": 1890, "height": 986 }, "focused": false, "focus": [ 13, 15 ], "border": "none", "current_border_width": 0, "layout": "splith", "orientation": "horizontal", "percent": null, "window_rect": { "x": 0, "y": 0, "width": 0, "height": 0 }, "deco_rect": { "x": 0, "y": 0, "width": 0, "height": 0 }, "geometry": { "x": 0, "y": 0, "width": 0, "height": 0 }, "window": null, "urgent": false, "marks": [ ], "fullscreen_mode": 1, "nodes": [ { "id": 13, "name": "*scratch*", "rect": { "x": 15, "y": 47, "width": 940, "height": 986 }, "focused": true, "focus": [ ], "border": "pixel", "current_border_width": 4, "layout": "none", "orientation": "none", "percent": 0.49735449735449733, "window_rect": { "x": 4, "y": 4, "width": 932, "height": 978 }, "deco_rect": { "x": 0, "y": 0, "width": 0, "height": 0 }, "geometry": { "x": 0, "y": 0, "width": 672, "height": 675 }, "window": 16777422, "urgent": false, "marks": [ ], "fullscreen_mode": 0, "nodes": [ ], "floating_nodes": [ ], "sticky": false, "type": "con", "pid": 8497, "app_id": null, "visible": true, "max_render_time": 0, "shell": "xwayland", "inhibit_idle": false, "idle_inhibitors": { "user": "none", "application": "none" }, "window_properties": { "class": "Emacs", "instance": "emacs", "title": "*scratch*", "transient_for": null, "window_type": "normal" } }, { "id": 15, "name": "guardian", "rect": { "x": 965, "y": 47, "width": 940, "height": 986 }, "focused": false, "focus": [ ], "border": "pixel", "current_border_width": 4, "layout": "none", "orientation": "none", "percent": 0.49735449735449733, "window_rect": { "x": 4, "y": 4, "width": 932, "height": 978 }, "deco_rect": { "x": 0, "y": 0, "width": 0, "height": 0 }, "geometry": { "x": 0, "y": 0, "width": 816, "height": 864 }, "window": 16777690, "urgent": false, "marks": [ ], "fullscreen_mode": 0, "nodes": [ ], "floating_nodes": [ ], "sticky": false, "type": "con", "pid": 8497, "app_id": null, "visible": true, "max_render_time": 0, "shell": "xwayland", "inhibit_idle": false, "idle_inhibitors": { "user": "none", "application": "none" }, "window_properties": { "class": "Emacs", "instance": "emacs", "title": "guardian", "transient_for": null, "window_type": "normal" } } ], "floating_nodes": [ ], "sticky": false, "num": 2, "output": "HDMI-A-1", "type": "workspace", "representation": "H[Emacs Emacs]" } }
00:03:21.256 [sway/desktop/transaction.c:414] Transaction 0x55bda3e39550 committing with 6 instructions
00:03:21.256 [sway/desktop/transaction.c:281] Applying transaction 0x55bda3e39550
00:03:21.256 [sway/ipc-server.c:960] Added IPC reply of type 0x0 to client 58 queue: [ { "success": true } ]
00:03:21.256 [DEBUG] [xwayland/xwm.c:968] XCB_PROPERTY_NOTIFY (924)
00:03:21.256 [DEBUG] [xwayland/xwm.c:1307] unhandled X11 event: 10
00:03:21.256 [DEBUG] [xwayland/xwm.c:968] XCB_PROPERTY_NOTIFY (924)
00:03:21.256 [DEBUG] [xwayland/xwm.c:968] XCB_PROPERTY_NOTIFY (924)
00:03:21.256 [DEBUG] [xwayland/xwm.c:1307] unhandled X11 event: 10
00:03:21.256 [DEBUG] [xwayland/xwm.c:968] XCB_PROPERTY_NOTIFY (924)
00:03:21.257 [sway/ipc-server.c:536] Client 24 writable
00:03:21.257 [sway/ipc-server.c:536] Client 58 writable
00:03:21.257 [sway/ipc-server.c:214] Client 58 hung up
00:03:21.257 [sway/ipc-server.c:566] IPC Client 58 disconnected
00:03:21.257 [sway/ipc-server.c:219] Client 41 readable
00:03:21.257 [sway/ipc-server.c:960] Added IPC reply of type 0x1 to client 41 queue: [ { "id": 14, "name": "2: <span font-family='Font Awesome 5 Free'><\/span>", "rect": { "x": 15, "y": 47, "width": 1890, "height": 986 }, "focus": [ 15, 13 ], "border": "none", "current_border_width": 0, "layout": "splith", "orientation": "horizontal", "percent": null, "window_rect": { "x": 0, "y": 0, "width": 0, "height": 0 }, "deco_rect": { "x": 0, "y": 0, "width": 0, "height": 0 }, "geometry": { "x": 0, "y": 0, "width": 0, "height": 0 }, "window": null, "urgent": false, "marks": [ ], "fullscreen_mode": 1, "nodes": [ ], "floating_nodes": [ ], "sticky": false, "num": 2, "output": "HDMI-A-1", "type": "workspace", "representation": "H[Emacs Emacs]", "focused": true, "visible": true }, { "id": 6, "name": "3: <span font-family='Font Awesome 5 Free'><\/span>", "rect": { "x": 15, "y": 47, "width": 1890, "height": 986 }, "focus": [ 5 ], "border": "none", "current_border_width": 0, "layout": "tabbed", "orientation": "none", "percent": null, "window_rect": { "x": 0, "y": 0, "width": 0, "height": 0 }, "deco_rect": { "x": 0, "y": 0, "width": 0, "height": 0 }, "geometry": { "x": 0, "y": 0, "width": 0, "height": 0 }, "window": null, "urgent": false, "marks": [ ], "fullscreen_mode": 1, "nodes": [ ], "floating_nodes": [ ], "sticky": false, "num": 3, "output": "HDMI-A-1", "type": "workspace", "representation": "T[Alacritty]", "focused": false, "visible": false } ]
00:03:21.257 [DEBUG] [xwayland/xwm.c:1307] unhandled X11 event: 10
00:03:21.257 [DEBUG] [xwayland/xwm.c:1307] unhandled X11 event: 10
00:03:21.257 [sway/ipc-server.c:536] Client 41 writable
00:03:21.257 [DEBUG] [xwayland/xwm.c:1307] unhandled X11 event: 10
00:03:21.257 [DEBUG] [xwayland/xwm.c:1307] unhandled X11 event: 10
00:03:21.257 [DEBUG] [xwayland/xwm.c:1307] unhandled X11 event: 10
00:03:21.257 [DEBUG] [xwayland/xwm.c:1307] unhandled X11 event: 10
00:03:21.257 [DEBUG] [xwayland/xwm.c:1307] unhandled X11 event: 10
...
```

- **Configuration File:**

Minimal sway configuration with only keyboard shortcuts for alacritty and bemenu.

- **Description:**

With two or more windows belonging to the same process, `swaymsg [instance=emacs] focus` or `swaymsg [instance=brave-browser] focus` leads to infinite xwm loop. All windows of the same process try to capture the focus infinitely. Thus, the process freezes.

`[xwayland/xwm.c:1307] unhandled X11 event: 10` is repeated for more than tens of thousands of times.
If windows are caught in an infinite loop, I have to kill them through `pkill -9`.

If multiple windows matching the focus criteria belong to different processes, this issue doesn't occur. This issue occurs only if two or more windows belong to the same process and are matched by focus criteria.

i3 deals with this issue by just focusing the last window among the multiple windows of the same process that match the focus criteria.

**Comment by Xyene** [comment]
Does this still happen with latest sway/wlroots master? https://github.com/swaywm/wlroots/pull/2355 was merged after 0.11.0, and fixes a very similar-sounding issue with Chrome.

**Comment by crocket** [comment]
I installed the latest commit of sway and wlroots. This issue is fixed on the latest commit.
